### DISCUSSION TITLE
"Cannot execute insert in a read-only transaction" with Function

### DISCUSSION THREAD

**Discussion by nruffilo** [supabase/supabase#1774] [Questions] [answered]
I'm using functions to simplify front-end code and ensure higher permissions.  For the most part they are working well, but I'm getting an odd error in one function:
""cannot execute INSERT in a read-only transaction"

The function in question is:
```sql
CREATE OR REPLACE FUNCTION GetHero()
RETURNS table(avatarurl varchar, displayname varchar, experiencepoints int4)
AS $Body$
  DECLARE userId uuid;
begin
  userId = auth.uid();
  IF NOT EXISTS (SELECT user_id FROM userdetail WHERE user_id = userId) THEN
    INSERT INTO userdetail
      (user_id, avatarurl, displayname, experiencepoints) VALUES (userId, '','New Hero',0);
  END IF;
  return 
    QUERY SELECT ud.avatarurl, ud.displayname, ud.experiencepoints FROM userdetail ud WHERE user_id = userId;
END
$Body$
LANGUAGE plpgsql VOLATILE;
```

Any thoughts?

**Comment by steve-chavez** [comment]
Hey @nruffilo,

How are you calling the `GetHero()` function?

**Reply by nruffilo** [reply]
I'm calling it using javascript - using supabase.rpc()

I use stored proceedures/functions throughout the app, some of them write to the DB, some read, all work fine, but this specific one gives me this error - but only when it needs to write to the userdetail table, if there is a record, it returns data correctly.

**Reply by steve-chavez** [reply]
Can you double check if the function is indeed VOLATILE? This error can happen when the function is IMMUTABLE or STABLE and it modifies the db.

Check with this query:

```sql
select provolatile from pg_proc where proname = 'GetHero'; 
-- or maybe select provolatile from pg_proc where proname = 'gethero'; 
-- it should show "v"
```

Also note that if your function has uppercase chars, it should be double quoted: `CREATE OR REPLACE FUNCTION "GetHero"`.

**Reply by nruffilo** [reply]
I prefer coding in camelcase, but I do realize that it makes everything lowercase.  When I ran the above query I got "v"

**Reply by nruffilo** [reply]
The issue was that my code was using .rpc("gethero").select("*")... and i needed to remove the .select("*") and it worked.  thank you for your help in getting me to think differently about the issue.

**Comment by nruffilo** [comment, marked as answer]
The issue was that my code was using .rpc("gethero").select("")... and i needed to remove the .select("") and it worked. thank you for your help in getting me to think differently about the issue.
